### PR TITLE
MR-11: employees

### LINKED ISSUE #11: Employees
- Invite user to a restaurant
- View employees/invitations
- View restaurants/invitations
- Update status by user
- Update status by restaurant

### PR DESCRIPTION
**Pull request by KirillSmirnow**
Closes #11

### CODE REVIEW

File: server/src/test/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeesRetrievalTest.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    @Test(expected = UserException::class)
+    fun `When get employees of non-existent restaurant, then failure`() {
+        employeeService.getEmployeesOfRestaurant(-200)
+    }
+
+    @TestConfiguration
+    @ComponentScan(
+        "org.itmodreamteam.myrest.server.service.employee",
+        "org.itmodreamteam.myrest.server.service.restaurant",
+        "org.itmodreamteam.myrest.server.service.user",
```
Comment (KirillSmirnow) [inline server/src/test/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeesRetrievalTest.kt:96]: As for unit testing it's better to test each Service in isolation from others by using mock objects

File: server/src/main/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeeServiceImpl.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            throw UserException("Сотрудник уже существует")
+        }
+        val employee = employeeRepository.save(
+            when (invitation.position) {
+                EmployeePosition.MANAGER -> Manager(restaurant, user)
+                EmployeePosition.WAITER -> Waiter(restaurant, user)
+            }
+        )
+        notificationService.notify(user, "${restaurant.name} хочет добавить вас в качестве сотрудника")
+        return getById(employee.id)
```
Comment (KirillSmirnow) [inline server/src/main/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeeServiceImpl.kt:45]: Use `toEmployeeInfo(employee)` as in other methods

File: server/src/main/kotlin/org/itmodreamteam/myrest/server/service/notification/FakeNotificationService.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import org.itmodreamteam.myrest.server.model.user.User
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+
+@Service
+class FakeNotificationService : NotificationService {
+
+    private val log = LoggerFactory.getLogger(javaClass)
+
+    override fun notify(user: User, text: String) {
```
Comment (Rznov) [inline server/src/main/kotlin/org/itmodreamteam/myrest/server/service/notification/FakeNotificationService.kt:12]: We really gonna pass User entity? We can use some of user dto's

**Review by Rznov — CHANGES_REQUESTED** [review]

File: server/src/main/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeeService.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import org.itmodreamteam.myrest.shared.restaurant.EmployeeInfo
+import org.itmodreamteam.myrest.shared.restaurant.EmployeeInvitation
+import org.itmodreamteam.myrest.shared.restaurant.EmployeeRestaurantStatus
+import org.itmodreamteam.myrest.shared.restaurant.EmployeeUserStatus
+
+interface EmployeeService {
+
+    fun inviteEmployee(restaurantId: Long, invitation: EmployeeInvitation): EmployeeInfo
+
+    fun updateEmployee(id: Long, userStatus: EmployeeUserStatus): EmployeeInfo
```
[the remark this answers is not in the document]
Comment (KirillSmirnow) [reply server/src/main/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeeService.kt:13]: That's a good question! One method is for changing Employee status by user him/her-self, another one is for changing it by the restaurant managers. The methods accept different parameters, update different Employee attributes, are invoked by different roles and, consequently, are going to have different permission checks in the upper layer.
It's better to implement authorization / permission checks in the upper layer (before invoking REST controller method) as SRP says.
Having a single method that behaves differently depending on current user would be less cleaner than having separate methods.

File: server/src/main/kotlin/org/itmodreamteam/myrest/server/service/notification/FakeNotificationService.kt (hunk 3 above)
Comment (KirillSmirnow) [reply server/src/main/kotlin/org/itmodreamteam/myrest/server/service/notification/FakeNotificationService.kt:12]: This method needs to know which user to notify. All we need to determine a user is its ID.
Here we pass User so as not to fetch it from DB inside this method once more.
If it is convenient, we can add `fun notify(userId: Long, text: String)` any time later.

**Comment by AlexIvchenko** [comment]
It's not related to code, but to model. I am not quite sure we need to have separate entities for Manager and Waiter for some reasons:
1. What if one employee can be both Manager and Waiter?
2. What happen if Waiter becomes Manager (and vise versa) at some time?

The basic answer for second question: remove from waiters table and insert into managers table. However, foreign keys and reference integrity are big deal (not for now, but for future features which will require reference to employee, e.g. waiter who is responsible to serve a table and who is needed to notify when a client wants to change coals in hookah). I think current model can produce logical issues which are difficult to fix. 

My proposal is to have single field List<Position> in Employee. Position can be just an enum {MANAGER, WAITER}.
It covers case when one employee can be both manager and waiter.
It is simple to make transitions between different positions.
No need to cast to specific classes to ensure position of employee.

**Review by Rznov — APPROVED** [review]

File: server/src/test/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeesRetrievalTest.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    lateinit var italianRestaurant: Restaurant
+    lateinit var frenchRestaurant: Restaurant
+
+    lateinit var italianRestaurantEmployees: List<User>
+    lateinit var frenchRestaurantEmployees: List<User>
+
+    @Before
+    fun setUp() {
+        italianRestaurant = restaurantRepository.save(Restaurant("Percorso", "Italian cuisine", "123"))
```
[the remark this answers is not in the document]
Comment (KirillSmirnow) [reply server/src/test/kotlin/org/itmodreamteam/myrest/server/service/employee/EmployeesRetrievalTest.kt:48]: Doing my best :)

**Pull request merged**